### ISSUE TITLE
yolov11OBB预测角度输出

### ISSUE THREAD

**Issue by wudiabin** [ultralytics/ultralytics#22920]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

yolov11旋转目标检测推理输出结果时，如果角度超过90°时如何处理。我在用X-AnyLabeling标注时，角度的计算方式为相比于水平方向顺时针旋转的角度，标注时一个物体相比于水平方向顺时针旋转156°，预测输出为66°左右，预测结果异常。貌似yolo内部推理时进行了相关的处理，我应该修改哪里保证让其显示原本的156°呢

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wudiabin, thank you for your interest in Ultralytics and for exploring YOLO11 OBB rotation angles 🚀  
This is an automated response to help you get support as quickly as possible — an Ultralytics engineer will review your issue and assist you in more detail soon 🙂  

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since your question is related to rotated object detection (OBB) and angle interpretation, it will be very helpful if you can provide more technical context so the team can reproduce and understand the behavior.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) including:
- The exact Ultralytics version (`ultralytics` pip version)
- A small snippet of your inference code (Python or CLI) that shows how you are running OBB prediction
- A minimal sample annotation from X-AnyLabeling (for example, a single object with the 156° angle you mentioned)
- The corresponding model config or model name you are using (for example, specific YOLO11 OBB model)
- A screenshot or visualization of the prediction vs. ground truth where 156° becomes about 66°

If this is a custom training ❓ Question about OBB angle conventions, please provide as much information as possible, including:
- How your labels are formatted/exported from X-AnyLabeling
- A few example images with their OBB annotations
- The exact command or code you used for training and prediction
- Any post-processing or visualization code you use to display angles  
and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Before the team investigates further, please ensure you have upgraded to the latest `ultralytics` package, including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8**, to verify the behavior still occurs in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by wudiabin** [comment]
Train_code:
```
import warnings
warnings.filterwarnings('ignore')
from ultralytics import YOLO

if __name__ == '__main__':
    # model = YOLO('ultralytics/cfg/models/11/yolo11s.yaml').load('yolo11s.pt')
    model = YOLO('yolo11s-obb.pt')  # loading pretrain weights
    model.train(data='datasets/20251121_datasets旋转目标检测/20251121_datasets/datasets.yaml',
                batch=8,
                epochs=200,        
                workers=2,
                device='0',
                imgsz=640,         
                optimizer='SGD',
                project='runs/train/20251121_obb',
                name='exp',
                task='obb',    
                patience=100,
                freeze=0,
                lr0=0.01,        
                degrees=0.0,         
                translate=0.1,     
                scale=0.5,         
                mosaic=0.1,        
                )
```


predict_code:

```
import warnings
warnings.filterwarnings('ignore')
from ultralytics import YOLO


if __name__ == '__main__':
    # model = YOLO(model='yolo11s.pt')
    # model = YOLO(model='yolo11s-obb.pt')
    model = YOLO(model='runs/train/20251121_obb/exp4/weights/best.pt')

    results = model.predict(source=r'datasets/20251121_datasets旋转目标检测/20251121_datasets/images/train',
                  imgsz=640,
                  device='0',
                  project='runs/obb',
                  task='obb',
                  name='exp',
                  save=True,
                  conf=0.7,
                  iou=0.3,
                )
```

**Comment by wudiabin** [comment]
X-AnyLabeling

<img width="1035" height="735" alt="Image" src="https://github.com/user-attachments/assets/85f51171-b93a-4085-a018-6acd36c4316b" />

label_txt:

<img width="1341" height="74" alt="Image" src="https://github.com/user-attachments/assets/516cc96f-d059-4895-b23b-d78bda9bc5cd" />

Result
left：label；right：predict

<img width="1781" height="721" alt="Image" src="https://github.com/user-attachments/assets/91eb580c-c5c6-4fd0-b317-5c3d8c57c1c7" />

It seems that the prediction results only go wrong when the angle exceeds 90 degrees. Maybe the angle is directly subtracted by 90 degrees, or perhaps the angle calculation should be changed to be based on the shorter side?

**Comment by wudiabin** [comment]
ultralytics/engine/results.py L572

<img width="773" height="770" alt="Image" src="https://github.com/user-attachments/assets/86ee10df-f0bb-4d6a-8517-3f08984383bc" />

**Comment by wudiabin** [comment]
My core question is, what is the angle range predicted by yolov11OBB? It seems to be only 0-90 degrees? In my experimental results, it seems that angles within 90 degrees are fine, but angles between 90-180 degrees have issues, with values deviating by 90 degrees, similar to 156-90=66 in the image above? Does yolov11OBB internally handle predicted angles over 90 degrees in a special way, and if so, where should I modify it to display the angle I want (156 degrees)? I need an angle representation range from 0 to 180 degrees.

**Comment by wudiabin** [comment]
ultralytics/nn/modules/head.py

<img width="1182" height="736" alt="Image" src="https://github.com/user-attachments/assets/9721c34e-4c13-4cb6-9dd6-5945443153d0" />

**Comment by Y-T-G** [comment]
Ultralytics OBB only supports angle between 0-90

**Comment by wudiabin** [comment]
> Ultralytics OBB only supports angle between 0-90

But I saw that in line 328 of the ultralytics/nn/modules/head.py file, the originally specified angle range is [-pi/4, 3pi/4], a 180-degree range. Shouldn't the predicted angles also have a 180-degree range?

<img width="1069" height="127" alt="Image" src="https://github.com/user-attachments/assets/0c5bea15-9541-4be5-bfab-c15d51f8fcbe" />

**Comment by Y-T-G** [comment]
It's 180 degrees to make the 0-90 not be at extreme ends, so that sigmoid multiplication can reach those values easily. But all the labels during training are converted to 0-90 using `cv2.minAreaRect`

**Comment by wudiabin** [comment]
> It's 180 degrees to make the 0-90 not be at extreme ends, so that sigmoid multiplication can reach those values easily. But all the labels during training are converted to 0-90 using `cv2.minAreaRect`

Thank you very much for your reply!
1. Does it mean that the angle range of [-pi/4, 3pi/4] is only to avoid having [0, pi/2] in an extreme position, rather than actually predicting an angle range of 180 degrees? Is my understanding correct?
2. If I want to output an angle range of [0, pi] in the prediction results, is that feasible? Does yolov11OBB support this, and if so, where should I make changes?

**Comment by Y-T-G** [comment]
1. Yes
2. It's not supported. You will have to change the dataloader code. YOLO OBB labels don't save angles. They only contain polygons which are converted to angles through minAreaRect.

**Comment by wudiabin** [comment]
> 1. Yes
> 2. It's not supported. You will have to change the dataloader code. YOLO OBB labels don't save angles. They only contain polygons which are converted to angles through minAreaRect.

Okay, got it. Thank you very much for your reply.

**Comment by wudiabin** [comment]
> 1. Yes
> 2. It's not supported. You will have to change the dataloader code. YOLO OBB labels don't save angles. They only contain polygons which are converted to angles through minAreaRect.

Finally, I just want to confirm again: the labels I marked only contain vertex coordinate information. In terms of the X-AnyLabeling standard, this cannot actually be passed to yolov11OBB. It only has (x1, y1), (x2, y2), (x3, y3), (x4, y4), and then yolov11OBB calculates a rotation angle between 0-90 degrees based on these coordinates and cv2.minAreaRect as the actual angle. The angle calculated in this way is not exactly the same as the angle I marked. Is my understanding correct?

**Comment by Y-T-G** [comment]
Yes
